library/data/bitvec.lean

```diff
@@ -74,92 +77,96 @@ end bitwise
-  | x y := prod.snd (sbb x y ff)
-
-  instance : has_zero (bitvec n) := ⟨zero⟩
-  /- Remark: we used to have the instance has_one only for (bitvec (succ n)).
-     Now, we define it for all n to make sure we don't have to solve unification problems such as:
-     succ ?n =?= (bit0 (bit1 one)) -/
-  instance : has_one (bitvec n) :=
-  match n with
-  | 0            := ⟨zero⟩
-  | (nat.succ n) := ⟨one⟩
-  end
+  def sbb (x y : bitvec n) (b : bool) : bool × bitvec n :=
+  let f := λ x y c, (bitvec.borrow x y c, bitvec.xor3 x y c) in
+  tuple.map_accumr₂ f x y b
+
+  def sub (x y : bitvec n) : bitvec n := prod.snd (sbb x y ff)
+
+  instance : has_zero (bitvec n) := ⟨bitvec.zero n⟩
+  instance : has_one (bitvec n)  := ⟨bitvec.one n⟩
   instance : has_add (bitvec n)  := ⟨add⟩
   instance : has_sub (bitvec n)  := ⟨sub⟩
   instance : has_neg (bitvec n)  := ⟨neg⟩
 
-  def mul : bitvec n → bitvec n → bitvec n
-  | x y :=
-    let f := λr b, cond b (r + r + y) (r + r) in
-    list.foldl f 0 (to_list x)
+  def mul (x y : bitvec n) : bitvec n :=
+  let f := λ r b, cond b (r + r + y) (r + r) in
+  list.foldl f 0 (to_list x)
 
   instance : has_mul (bitvec n)  := ⟨mul⟩
 end arith
 
 section comparison
   variable {n : ℕ}
 
-  def uborrow : bitvec n → bitvec n → bool := λx y, prod.fst (sbb x y ff)
+  def uborrow (x y : bitvec n) : bool := prod.fst (sbb x y ff)
 
-  def ult (x y : bitvec n) : Prop := uborrow x y = tt
+  def ult (x y : bitvec n) : Prop := uborrow x y
   def ugt (x y : bitvec n) : Prop := ult y x
 
   def ule (x y : bitvec n) : Prop := ¬ (ult y x)
-  def uge : bitvec n → bitvec n → Prop := λx y, ule y x
+  def uge (x y : bitvec n) : Prop := ule y x
 
-  def sborrow : bitvec (succ n) → bitvec (succ n) → bool := λx y,
+  def sborrow : Π {n : ℕ}, bitvec n → bitvec n → bool
+  | 0        _ _ := ff
+  | (succ n) x y :=
      bool.cases_on
        (head x)
        (bool.cases_on (head y) (uborrow (tail x) (tail y)) tt)
        (bool.cases_on (head y) ff (uborrow (tail x) (tail y)))
 
-  def slt : bitvec (succ n) → bitvec (succ n) → Prop := λx y,
-     sborrow x y = tt
-
-  def sgt : bitvec (succ n) → bitvec (succ n) → Prop := λx y, slt y x
-  def sle : bitvec (succ n) → bitvec (succ n) → Prop := λx y, ¬ (slt y x)
-  def sge : bitvec (succ n) → bitvec (succ n) → Prop := λx y, sle y x
+  def slt (x y : bitvec n) : Prop := sborrow x y
+  def sgt (x y : bitvec n) : Prop := slt y x
+  def sle (x y : bitvec n) : Prop := ¬ (slt y x)
+  def sge (x y : bitvec n) : Prop := sle y x
 
 end comparison
 
-section from_bitvec
+section conversion
   variable {α : Type}
 
-  -- Convert a bitvector to another number.
-  def from_bitvec [has_add α] [has_zero α] [has_one α] {n : nat} (v : bitvec n) : α :=
-  let f : α → bool → α := λr b, cond b (r + r + 1) (r + r) in
-  list.foldl f (0 : α) (to_list v)
-end from_bitvec
+  protected def ofNat : Π (n : ℕ), nat → bitvec n
+  | 0        x := nil
+  | (succ n) x := ofNat n (x / 2) ++ₜ [to_bool (x % 2 = 1)]
+
```

Comment:
I was just about to comment on this :) . I deliberately decided to go against the convention in my own project for these definitions, because having functions containing `of_` is just confusing when proving things about them. With that being said, now is probably not the time to reopen that discussion, so I'm changing it right now.